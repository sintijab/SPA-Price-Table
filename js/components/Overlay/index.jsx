import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { Input, ButtonBase } from '@material-ui/core';
import DOMPurify from 'dompurify';
import { updateData } from '../../utils/functions';
import { validatePrice } from '../../utils/priceFunctions';
import { URL_HOST, ALL_COMPANIES, BUDGET } from '../../constants/endpoints';
import {
  CLOSE_BUTTON,
  INPUT_ERROR_TEXT,
  INPUT_VALID_TEXT,
} from '../../constants/titles';
import styles from './index.scss';

const Overlay = ({ item, hideModal, updateBudget }) => {
  const {
    register,
    errors,
    handleSubmit,
  } = useForm({
    mode: 'onChange',
    reValidateMode: 'onChange',
  });
  const [infoMessage, setAdditionalInfo] = useState(false);
  const {
    name,
    budget,
    id,
  } = item || {};
  const budgetTotal = validatePrice(budget);
  const onSubmit = (e) => {
    const params = {
      url: `${URL_HOST}${ALL_COMPANIES}/${id}`,
      property: BUDGET,
      value: e.totalBudget,
    };
    updateData(params)
      .then(() => {
        setAdditionalInfo(INPUT_VALID_TEXT);
        updateBudget(id, e.totalBudget);
      }).catch((err) => {
        console.error(err);
      });
  };
  const closeOverlay = () => {
    setAdditionalInfo(null);
    hideModal();
  };
  const buttonText = `${errors.totalBudget ? 'Disabled' : 'Submit'}`;
  const { sanitize } = DOMPurify;
  return (
    <div className={styles.modal}>
      <div className="close-btn" role="presentation" onClick={() => closeOverlay()}>
        {CLOSE_BUTTON}
      </div>
      <div
        className={styles.modal_item}
        // eslint-disable-next-line
        dangerouslySetInnerHTML={{ __html: sanitize(`<h3>${name}</h3>`) }}
      />
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          name="totalBudget"
          type="number"
          inputProps={{
            step: 'any',
          }}
          defaultValue={budgetTotal}
          onChange={() => setAdditionalInfo(null)}
          label="total price"
          inputRef={register({
            required: true,
            maxLength: 20,
            pattern: /^[0-9]+(\.[0-9][0-9])?$/,
          })}
          error={!!errors.totalBudget}
        />
        <br />
        {errors.totalBudget && (
          <h4 className="error-text">{INPUT_ERROR_TEXT}</h4>
        )}
        {!!infoMessage && (
          <h4 className="valid-text">{infoMessage}</h4>
        )}
        <ButtonBase
          type="submit"
          disabled={!!errors.totalBudget}
          classes={{ root: 'button' }}
        >
          {buttonText}
        </ButtonBase>
      </form>
    </div>
  );
};

Overlay.propTypes = {
  item: PropTypes.objectOf(PropTypes.oneOfType(
    [PropTypes.string, PropTypes.number],
  )),
  hideModal: PropTypes.func,
  updateBudget: PropTypes.func,
};

Overlay.defaultProps = {
  item: {},
  hideModal: () => {},
  updateBudget: () => {},
};

export default Overlay;
