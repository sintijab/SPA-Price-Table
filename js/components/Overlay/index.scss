@import '../../../styles/main';

.modal {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: 1rem;
  padding: 10rem;
  border-radius: 1rem;
  line-height: 2;
  background: $white;
}

.modal_item {
  display: block;
}
