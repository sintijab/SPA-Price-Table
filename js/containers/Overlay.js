import { connect } from 'react-redux';
import { updateTotalBudget } from '../actions/priceActions';
import { closeModal } from '../actions/modalActions';
import Overlay from '../components/Overlay';


const mapDispatchToProps = (dispatch) => ({
  hideModal: () => dispatch(closeModal()),
  updateBudget: (id, price) => dispatch(updateTotalBudget(id, price)),
});

export default connect(
  null,
  mapDispatchToProps,
)(Overlay);
