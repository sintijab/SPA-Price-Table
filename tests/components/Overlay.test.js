import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Input, ButtonBase } from '@material-ui/core';
import toJson from 'enzyme-to-json';
import Overlay from '../../js/components/Overlay';

Enzyme.configure({ adapter: new Adapter() });

describe('<Overlay />', () => {
  let props = {
    id: 1,
    name: "MartinFirma",
    budget: 65345664332234,
    budget_spent: 45000,
    date_of_first_purchase: "2119-07-07"
  };

  it('renders <Overlay /> component', () => {
    const wrapper = shallow(<Overlay item={props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Input field', () => {
    const wrapper = shallow(<Overlay item={props} />);
    const input = wrapper.find(Input).at(0);
    expect(input).toHaveLength(1);
  });

  it('calls hideModal action by closing overlay', () => {
    const spy = jest.fn();
    const wrapper = mount(<Overlay item={props} hideModal={spy}/>);
    wrapper.find(".close-btn").simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('validates prefilled Input value with valid integer number', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: 123 }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('123.00');
  });

  it('validates prefilled Input value with string numeric value', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: '1254.489999993' }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('1254.49');
  });

  it('validates prefilled Input value with valid floating point number', () => {
    const wrapper = shallow(<Overlay  />);
    wrapper.setProps({ item: { budget: 123.123 }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('123.12');
  });

  it('validates prefilled Input value with not valid value', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: '123.ss7888[/.123' }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe(null);
  });

  it('outputs correct prefilled Input values with floating-point exceptions', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: 1.015 }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('1.02');
  });

  it('outputs correct prefilled Input price values with floating-point exceptions', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: 1.0654 }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('1.07');
  });

  it('outputs correct prefilled Input price values with floating-point exceptions', () => {
    const wrapper = shallow(<Overlay />);
    wrapper.setProps({ item: { budget: 1.27499 }});
    expect(wrapper.find(Input).at(0).props().defaultValue).toBe('1.27');
  });
});
