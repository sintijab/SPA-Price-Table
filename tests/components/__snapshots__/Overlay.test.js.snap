// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> renders <Overlay /> component 1`] = `
<div
  className="modal"
>
  <div
    className="close-btn"
    onClick={[Function]}
    role="presentation"
  >
    Close
  </div>
  <div
    className="modal_item"
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h3>MartinFirma</h3>",
      }
    }
  />
  <form
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(Input))
      defaultValue="65345664332234.00"
      error={false}
      inputProps={
        Object {
          "step": "any",
        }
      }
      inputRef={[Function]}
      label="total price"
      name="totalBudget"
      onChange={[Function]}
      type="number"
    />
    <br />
    <WithStyles(ForwardRef(ButtonBase))
      classes={
        Object {
          "root": "button",
        }
      }
      disabled={false}
      type="submit"
    >
      Submit
    </WithStyles(ForwardRef(ButtonBase))>
  </form>
</div>
`;
